### PR TITLE
[VAX-532] Introduce dependency info for row add-wins semantics

### PR DESCRIPTION
**Pull request by define-null**
Tags mechanism is added to Satellite protocol to track row dependencies in the form similar to ORSet, to provide add-wins semantics for concurrent operations. Tags stored with the row determine wherever or not row exists.

### CODE REVIEW

File: src/satellite/process.ts
Comment (icehaunter) [inline src/satellite/process.ts:737]: Can you please elaborate a little on why this is valid? I'm not sure I follow logic of running a bulk delete operation when applying a transaction, and also limiting the delete operation to the transactions in the same microsecond.

File: src/satellite/process.ts
Comment (icehaunter) [inline src/satellite/process.ts:335]: So as far as I can tell, `shadowEntries` is cached only in the scope of this function call, i.e. when the same row is touched multiple times in the same oplog "span", right?

File: src/satellite/process.ts
Comment (icehaunter) [inline src/satellite/process.ts:599]: If `oplog` can be undefined, let's reflect that in the types of the argument (here and up the call chain). Otherwise, let's drop the condition

**Review by icehaunter — CHANGES_REQUESTED** [review]
Looks cool! I'm still wrapping my head around some shadow entry logic. Meanwhile I had some minor comments and questions

File: src/satellite/client.ts
Comment (define-null) [reply src/satellite/client.ts:638]: Good question. So on a direction from Satellite, origin is the same as an id which is passed in the Auth message. So it might very well be ignored by the server. However on the way back it is important to specify that, so that Satellite knows, when it receives transaction which originated from it.

File: src/satellite/process.ts
Comment (define-null) [reply src/satellite/process.ts:510]: If you don't mind I'm gonna clean it up once we are happy with the review. I left it in case we will need more unit tests. I will wait till I get approval from @balegas on that one.

File: src/satellite/process.ts
Comment (define-null) [reply src/satellite/process.ts:343]: same as above

File: src/satellite/process.ts
Comment (define-null) [reply src/satellite/process.ts:737]: I should probably add a comment for that in the code but current logic is - to clean oplog when we receive transaction from Electric that corresponds to this part of oplog. Previously we would do such a GC when receiving ack.

File: src/satellite/process.ts
Comment (define-null) [reply src/satellite/process.ts:335]: Yes, that's the idea.

File: src/satellite/process.ts
Comment (icehaunter) [reply src/satellite/process.ts:737]: I'd really appreciate a comment in code about it with this, and probably an explanation as to why is exact equality of items in the oplog on a millisecond time precision is a good indicator of garbage, for posterity

File: src/satellite/merge.ts
Comment (balegas) [reply src/satellite/merge.ts:61]: We discussed this and I remember agreeing with code, but need to refresh my memory and have things explained to me from your perspective.

When applying incoming operations, we want to add to the local shadow table new incoming tags and make sure that we don't re-add tags that have been cleared by local operations (this is to ensure that a local DELETE after a local INSERT that is being re-deliverred does not re-introduce new tags). Is that what you're doing here?

I would expect that the merge of two entries would generate a new entry with tag being the the union of local and remote tag minus the local clear tags (in this case tag, would have to be a set --- I think this is a valid case: local insert with remote insert). 

ClearTags, I think can always be an union, because they just remove stuff that has been seen.

Somehow, it seems you're doing something different and I need to understand better.

Comment (balegas) [reply src/satellite/merge.ts:61]: I think this is the most critical part of this PR and I want to stop review here before continuing, so let me also make another comment: I think we need more unit tests to test more isolated bits of the code. Current tests are hard to follow for someone coming in. I have presented at least a couple of counter examples, if I am correct in some of my examples, we should have unit tests for showing the issue and use that as inspiration to make more tests.

File: src/satellite/process.ts
Comment (balegas) [inline src/satellite/process.ts:725]: Somehow I think you're doing the right thing, but I do not understand why this is called clear_tags and is making a lot of confusion to me (as of my previous long comment). Let's discuss this again.

**Review by balegas — CHANGES_REQUESTED** [review]
Minor comments and need to understand better your implementation of entries merging. I suppose this was a long time ago for me too :).

As a general comment, we need to have more unit tests for specific parts of the code, instead of testing everything top down

File: src/satellite/oplog.ts
Comment (balegas) [reply src/satellite/oplog.ts:230]: After finishing the review, I imagine this might be right, but I just don't understand really the meaning of clear tags.

File: src/satellite/oplog.ts
Comment (define-null) [reply src/satellite/oplog.ts:136]: I'm not sure I follow what you mean

Comment (balegas) [reply src/satellite/oplog.ts:136]: I wonder if you could change the interface to not receive the lambda for generating the tag.

File: src/satellite/process.ts
Comment (define-null) [reply src/satellite/process.ts:776]: I left a comment to have an opportunity to discuss this further. 

With current version of GC we will keep oplog entries for longer time frames then we used to have, 
which will causes extra performance penalty when syncing transactions from Electric, and extra space
consumed on the device as we do not merge entries in the oplog, and they going to be GCed only on 
receiving same transaction from Electric and not just when Electric have confirmed that transaction has
been written. Also sending the whole transaction back to the source looks like an overkill in this situation,
 so it's also something for us to consider for optimization.

File: src/satellite/oplog.ts
Comment (define-null) [reply src/satellite/oplog.ts:136]: Lambda is used to capture the origin field, which is not present in the oplogEntry

File: src/satellite/client.ts
Comment (define-null) [reply src/satellite/client.ts:638]: The comment in proto file explicitly says about it:
```
   // Globally unique id of the source that transaction originated from. For
    // data coming from Satellite this field is ignored. For data coming from
    // Electric this field can be used to deduce if the incoming transaction
    // originated on this Satellite instance or not.
```
Let me know if further clarification on the protocol is desired

File: src/satellite/process.ts
Comment (define-null) [reply src/satellite/process.ts:889]: > but we could piggyback the ack with the transaction message.
I'm not sure what you mean here. Ack to the transaction and sending transaction back are decoupled processes and the underlying system doesn't have matching between lsn and transaction.

> Second, I think it looks unsafe to rely on timestamp,
As long as timestamp is strictly monotonically increasing for the target it should be safe. If it doesn't - then conflict resolution wouldn't work properly. We can store mapping `transaction -> lsn position` on Satellite if you think that's more appropriate.

File: src/satellite/merge.ts
Comment (define-null) [reply src/satellite/merge.ts:44]: I think merging opType and ChangesOpType requires a bit of extra work which is not really relevant for this PR.
They are used for different purposes - for oplog entries and for merging.

File: src/satellite/oplog.ts
Comment (balegas) [reply src/satellite/oplog.ts:136]: I'm sorry to insist, but the param for getTag is a Date object from entry.timestamp in every reference to getTag that I could find.

Comment (define-null) [reply src/satellite/oplog.ts]: Now I see what you meant, fixed.

File: src/satellite/process.ts
Comment (balegas) [reply src/satellite/process.ts]: >  Also sending the whole transaction back to the source looks like an overkill in this situation,
so it's also something for us to consider for optimization.
We could just send a gist of the transaction and not the transaction itself, but we need to verify the correctness of that. Right now, the transaction that we receive might not be exactly the one that we've sent (it will be the result of applying it on the server against concurrent outstanding operations). We would have to ensure that there are no corner cases in skipping the transaction... I am not very concerned with avoiding transmitting the transaction, because we have to send it to all other satellites anyways.

> With current version of GC we will keep oplog entries for longer time frames
We can eventually compact the log by merging entries that are safe to merge and other strategies. Compaction would be very efficient at least in the worst case for growing the oplog, which is when the client is disconnected.

I hope that the oplog doesn't grow so large when the user is online, otherwise it will not be a very 'real-time' system. In that sense I believe that the cost of the oplog should be much smaller compared to the cost of keeping application data on the device.

File: src/satellite/process.ts
Comment (balegas) [reply src/satellite/process.ts:889]: >  Ack to the transaction and sending transaction back are decoupled processes and the underlying system doesn't have matching between lsn and transaction.

In the new approach, I think there could be no explicit ack message, because it is implied by receiving the message from the same origin. We would need to have a way to associate that with LSN, but this could be done by extending the transaction message to contain an optional ack field with the corresponding lsn position (which I don't know, but suppose is available on the server).

File: src/satellite/oplog.ts
Comment (define-null) [reply src/satellite/oplog.ts]: You are correct that we should not have a special case for delete here, I confused this with the scenario when data is coming from Electric, as there OPTYPES.delete means no tags. As for the entry.timestamp this tag is implicit based on the transaction info, so we optionally generate it when necessary

File: src/satellite/merge.ts
Comment (define-null) [reply src/satellite/merge.ts:47]: PR was implemented under assumption that we never reapply incoming transactions, that was also motivation for doing GC the way we do. If we need to reapply as you pointed out in another comment (which it seems we have to) then this logic is wrong.

Comment (define-null) [reply src/satellite/merge.ts:48]: Removed this logic, as it's no longer valid for tags approach.

File: src/satellite/process.ts
Comment (balegas) [inline src/satellite/process.ts:720]: I still find it confusing that this is called clearTags because what determines the optype is the presence of shadow tag and why in the `apply()` we also set tags with clearTags. 

We need to test this functions in isolation please. Could you make a simpler version of some of your tests that just generates the entries, calls apply, and checks what is the vicinity of `'INSERT wins over DELETE and restored deleted values'

File: src/satellite/process.ts
Comment (balegas) [inline src/satellite/process.ts:350]: I think this is okay, but let's just discuss to make sure we understand:
- The case when we end up with the tag in the clear set and in the tag set is INSERT.INSERT
- Is it a problem to have the tag in clear and tag sets? No, if when applying an incoming operation we first remove the clear set and then add the tag.

Having said that, I don't disagree with this code.

File: src/satellite/process.ts
Comment (balegas) [inline src/satellite/process.ts:28]: In this function, couldn't we take advantage of the source to fix the issue with timestamp comparison?
(it seems it still has the issue when timestamps are equal)

File: src/satellite/merge.ts
Comment (balegas) [inline src/satellite/merge.ts:44]: No longer used

File: src/satellite/oplog.ts
Comment (balegas) [inline src/satellite/oplog.ts:167]: If I understand correctly, when we compute a snapshot, we compute the clear set per entry and here we are taking the latest clear set from all entries.
I think we could have the following case:
t=[1], c=[]
t=[], c=[1]
t=[2], c=[]

And we would lose the clear of tag 1. Please provide a counter example if this is not a correct interpretation.

Comment (define-null) [reply src/satellite/oplog.ts]: fixed

File: src/satellite/process.ts
Comment (define-null) [reply src/satellite/process.ts:28]: yes, that was the plan originally, fixed

File: test/support/satellite-helpers.ts
Comment (define-null) [inline test/support/satellite-helpers.ts:94]:
```suggestion
```

File: src/satellite/process.ts
Comment (define-null) [inline src/satellite/process.ts:345]:
```suggestion
```

File: src/satellite/process.ts
Comment (define-null) [inline src/satellite/process.ts:371]:
```suggestion
```

File: src/satellite/process.ts
Comment (define-null) [inline src/satellite/process.ts:506]:
```suggestion
```

File: src/satellite/process.ts
Comment (define-null) [inline src/satellite/process.ts:529]:
```suggestion
```

File: src/satellite/process.ts
Comment (define-null) [inline src/satellite/process.ts:730]:
```suggestion
```

File: test/satellite/process.tags.test.ts
Comment (balegas) [inline test/satellite/process.tags.test.ts:149]: Please remove the long explanation (which was meant to you) and adapt the test name accordingly with other tests you created.

File: test/satellite/process.tags.test.ts
Comment (balegas) [inline test/satellite/process.tags.test.ts:260]: This and the following tests are not necessary since the intent was just to show the issue would not occur in scenarios with just two ops.

File: src/satellite/oplog.ts
Comment (balegas) [inline src/satellite/oplog.ts:174]: Changes/TableChanges should be consistent with type name

File: src/satellite/merge.ts
Comment (balegas) [inline src/satellite/merge.ts:12]: watch for naming convention

**Review by balegas — APPROVED** [review]
I think it looks goods.

**Comment by define-null** [comment]
Rebased on latest main.

File: test/satellite/process.tags.test.ts
Comment (define-null) [reply test/satellite/process.tags.test.ts]: removed

File: test/satellite/process.tags.test.ts
Comment (define-null) [reply test/satellite/process.tags.test.ts]: removed

**Review by balegas — APPROVED** [review]
Looks good to me. Great work.
Please close VAX-391 too.

**Pull request merged**
